src/sciline/pipeline.py

```diff
@@ -74,14 +94,139 @@ def _bind_free_typevars(tp: TypeVar | Key, bound: Dict[TypeVar, Key]) -> Key:
         return tp
 
 
+def _find_all_paths(
+    graph: Mapping[T, Collection[T]], start: T, end: T
+) -> List[List[T]]:
+    """Find all paths from start to end in a DAG."""
+    if start == end:
+        return [[start]]
+    if start not in graph:
+        return []
+    paths = []
+    for node in graph[start]:
+        for path in _find_all_paths(graph, node, end):
+            paths.append([start] + path)
+    return paths
+
+
+def _find_nodes_in_paths(
+    graph: Mapping[T, Tuple[Callable[..., Any], Collection[T]]], start: T, end: T
```

Comment:
This function and `_find_all_paths` use different graph types. Can you give the arguments a clearer name to distinguish? The type hint alone doesn't really say what kind of graph we are dealing with.

Also, why are there different graph types in the first place?